pyop2/sparsity.pyx

```diff
@@ -56,304 +54,119 @@ cdef extern from "petsc.h":
                                  PetscScalar*, PetscInsertMode)
     int MatSetValuesLocal(PETSc.PetscMat, PetscInt, PetscInt*, PetscInt, PetscInt*,
                           PetscScalar*, PetscInsertMode)
+    int MatPreallocatorPreallocate(PETSc.PetscMat, PetscBool, PETSc.PetscMat)
+    int MatXAIJSetPreallocation(PETSc.PetscMat, PetscInt, const PetscInt[], const PetscInt[],
+                                const PetscInt[], const PetscInt[])
 
+cdef extern from "petsc/private/matimpl.h":
+    struct _p_Mat:
+        void *data
```

Comment:
This is how the PETSc runtime delegation pattern is implemented.  See https://github.com/firedrakeproject/petsc/blob/firedrake/include/petsc/private/matimpl.h#L375